src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -52,107 +54,240 @@ final class P11SecretKeyFactory extends SecretKeyFactorySpi {
     // algorithm name
     private final String algorithm;
 
+    // PBEKeyInfo if algorithm is PBE
+    private final PBEKeyInfo svcPbeKi;
+
     P11SecretKeyFactory(Token token, String algorithm) {
         super();
         this.token = token;
         this.algorithm = algorithm;
+        this.svcPbeKi = getPBEKeyInfo(algorithm);
     }
 
-    private static final Map<String,Long> keyTypes;
+    private static final Map<String, KeyInfo> keyInfo = new HashMap<>();
+    private static final KeyInfo HMAC = new KeyInfo("HMAC", PCKK_HMAC);
+    private static final KeyInfo SSLMAC = new KeyInfo("SSLMAC", PCKK_SSLMAC);
+
+    static KeyInfo getKeyInfo(String algo) {
+        KeyInfo ki = keyInfo.get(algo);
+        if (ki == null) {
+            String algoUpper = algo.toUpperCase(Locale.ENGLISH);
+            ki = keyInfo.get(algoUpper);
+            if (ki == null) {
+                if (algoUpper.startsWith("HMAC")) {
+                    return HMAC;
+                } else if (algoUpper.startsWith("SSLMAC")) {
+                    return SSLMAC;
+                }
+            }
+        }
+        return ki;
+    }
+
+    static PBEKeyInfo getPBEKeyInfo(String algo) {
+        if (getKeyInfo(algo) instanceof PBEKeyInfo pbeKi) {
+            return pbeKi;
+        }
+        return null;
+    }
+
+    private static void putKeyInfo(KeyInfo ki) {
+        keyInfo.put(ki.algo, ki);
+        keyInfo.put(ki.algo.toUpperCase(Locale.ENGLISH), ki);
+    }
+
+    static sealed class KeyInfo permits PBEKeyInfo {
+        public final String algo;
+        public final long keyType;
+
+        KeyInfo(String algo, long keyType) {
+            this.algo = algo;
+            this.keyType = keyType;
+        }
+
+        static boolean checkUse(KeyInfo ki, KeyInfo si) {
+            if (si instanceof PBEKeyInfo && !si.algo.equals(ki.algo)) {
+                // PBE services require a PBE key of the same algorithm.
+                return false;
+            }
+            if (ki instanceof PBKDF2KeyInfo) {
+                // We cannot tell what the PBE key was derived for,
+                // so any service is allowed in principle.
+                return true;
+            }
+            return ki.keyType == si.keyType;
```

Comment:
So, for non-PBE key info, algos do not have to match?

Reply:
For services and keys cases in which algorithms identity-match —irrespective if they are PBE or non-PBE—, KeyInfo::checkUse is not called and execution moves forward as if the check passed (see [here](https://github.com/openjdk/jdk/blob/ab7ffd56bb8b93d513023d0136df55a6375c3286/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java#L285)). The same is true for services that accept any key type, such as those whose pseudo types are PCKK_HMAC or PCKK_SSLMAC.

The ki.keyType == si.keyType success value affects cases in which algorithms are different but it's still possible to use the key in the service. One example that would hit this path is a PBE key derived for AES that it's used in an AES Cipher service. For non-PBE keys and services cases, one example is algorithms "RC4" and "ARCFOUR" that have both the underlying CKK_RC4 key type. Notice that this latter case is not new: previous to this enhancement proposal, key types were compared as well (see [here](https://github.com/openjdk/jdk/blob/jdk-21%2B14/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java#L147)).

For non-PBE keys and services cases, what is new with this enhancement is to accept them if their algorithms are identity-equal. This condition necessarily means that key types are equal —the opposite is obviously not true—. One minor detail, when we refer to the algorithms equality trivial pass condition, it's an object identity comparison for performance.

Reply:
As commented [here](https://github.com/openjdk/jdk/pull/12396#discussion_r1142245451), I'll add a comment to the code explaining this execution path.